src/swoop/cache/hash_main.py

```diff
@@ -0,0 +1,277 @@
+    for arg in args:
+        sha.update(str(arg).encode())
+    hashed = base64.b64encode(sha.digest())
+    return hashed
+
+
+def traverse(dict_or_list: Union[Dict, List], path: List = []):
+    """
+    Traverses the entire payload object to yield a list of paths to all nodes.
+
+    Parameters:
+            payload (Dict): The input payload whose keys will be ordered
+                            alphabetically.
+
+    Returns:
+            result (Dict): The input payload with keys (including nested)
+                           ordered alphabetically.
+    """
+    if isinstance(dict_or_list, dict):
+        iterator = sorted(dict_or_list.items())
+    else:
+        iterator = enumerate(dict_or_list)
+    for k, v in iterator:
+        yield path + [k], v,
+        if isinstance(v, (dict, list)):
+            for k, v in traverse(v, path + [k]):
+                yield k, v
+
+
+def order_dict(payload: Dict) -> Dict:
+    """
+    Reorders the keys in the payload object alphabetically.
+
+    Parameters:
+            payload (Dict): The input payload whose keys will be ordered
+                            alphabetically.
+
+    Returns:
+            result (Dict): The input payload with keys (including nested)
+                           ordered alphabetically.
+    """
+    result = {}
+    for k, v in sorted(payload.items()):
+        if isinstance(v, dict):
+            result[k] = order_dict(v)
+        elif isinstance(v, list):
+            result[k] = []
+            for i in range(len(v)):
+                elem = v[i]
+                if isinstance(elem, dict):
+                    result[k].append(order_dict(elem))
+                else:
+                    result[k].append(elem)
+        else:
+            result[k] = v
+
+    return result
+
+
+def eval_to_list(expressions: List[str], path_list: List[str]) -> List[str]:
+    """
+    Finds the paths in the path list that match the expressions in
+    either the includes or excludes lists using regex.
+
+    Parameters:
+            expressions (List[str]): A list of expressions written in dot notation
+                that we want to evaluate
+            path_list (List[str]): A list of all paths to all nodes in the input
+                                   payload
+
+    Returns:
+            match_list (str): A list of paths in the path_list that match
+            with the expressions in the expressions list.
+    """
+
+    match_list = []
+
+    for exp in expressions:
+        regex = ""
+        split_exp = exp.split(".")
+        for e in split_exp:
+            if "[" in e:
+                sub = e.split("[")
+                if "[*]" in e:
+                    regex += sub[0] + ".[0-9]."
+                else:
+                    regex += sub[0] + ".[" + sub[1] + "."
+            elif e == "*":
+                regex += ".[-a-z0-9]."
+            else:
+                regex += e + "."
+        regex = regex.strip(".")
+        r = re.compile(regex)
```

Comment:
I think the problem with reducing things like index selection to strings makes optimizing away the need to compile and match a regex for each pattern really hard. Ideally you would only need a single regex, and indeed I think you could rather easily work out a way to do this for key names, but my idea breaks down when trying to work out how best to accommodate the possibility of ranges in the index queries.

Probably a bit because such a problem is hard, and a bit because reducing such a query to a regex is likely impossible. Consider the case of `[:-1]`...